### PR TITLE
[Neo Store UT] Add leveldb thread UT and readme.

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

Leveldb itself is not thread safe, thus having this pr to demonstrate how leveldb snapshot behaves under multi thread env. 

Quote from leveldb document https://github.com/google/leveldb/blob/main/doc/index.md :

> A database may only be opened by one process at a time. The leveldb implementation acquires a lock from the operating system to prevent misuse. Within a single process, the same leveldb::DB object may be safely shared by multiple concurrent threads. I.e., different threads may write into or fetch iterators or call Get on the same database without any external synchronization (the leveldb implementation will automatically do the required synchronization). **However other objects (like Iterator and WriteBatch) may require external synchronization.** If two threads share such an object, they must protect access to it using their own locking protocol. More details are available in the public header files.

Fixes # (issue)

## Type of change

- [ ] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] TestMultiThreadLevelDbSnapshotPut
- [x] TestMultiThreadLevelDbSnapshotPutWithoutCommit
- [x] TestMultiThreadLevelDbSnapshotPutWithLocker
- [x] TestOneSnapshotPerThreadLevelDbSnapshotPut

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by cschuchardt88 — COMMENTED** [review]
But `leveldb` does support multi-threading and so does the `interface` (`IStore`). The classes that inherit `IStore` are not thread safe. I said this at the begin when i started work here.

**Comment by Jim8y** [comment]
> But `leveldb` does support multi-threading and so does the `interface` (`IStore`). The classes that inherit `IStore` are not thread safe. I said this at the begin when i started work here.

No, it does not. Did i do anything saying it does here? Aren't we are saying exactly the same thing?

**Comment by cschuchardt88** [comment]
They you think your tests are failing than? Your testing the wrong thing, you need to be testing `SnapshotCache` and `DataCache`

**Comment by Jim8y** [comment]
> Then why you think your tests are failing than? Your testing the wrong thing, you need to be testing `SnapshotCache` and `DataCache`

which test are failing, aren't they all passed?

**Comment by cschuchardt88** [comment]
> > Then why you think your tests are failing than? Your testing the wrong thing, you need to be testing `SnapshotCache` and `DataCache`
> 
> which test are failing, aren't they all passed? Will test snapshotcache later.

image

**Comment by Jim8y** [comment]
> > > Then why you think your tests are failing than? Your testing the wrong thing, you need to be testing `SnapshotCache` and `DataCache`
> > 
> > 
> > which test are failing, aren't they all passed? Will test snapshotcache later.
> 
> image

Will update.

File: tests/Neo.Plugins.Storage.Tests/StoreTest.MultiThread.cs [hunk 1]
```diff
@@ -12,101 +12,157 @@
...
-                snapshot.Put(testKey, value);
-                snapshot.Commit();
-            });
+                for (var i = 0; i < tasks.Length; i++)
+                {
+                    var value = new byte[] { 0x04, 0x05, 0x06, (byte)i };
+                    tasks[i] = Task.Run(() =>
+                    {
+                        // Introduce delay to increase conflict chance
+                        Thread.Sleep(new Random().Next(1, 10));
```
Comment (cschuchardt88) [inline tests/Neo.Plugins.Storage.Tests/StoreTest.MultiThread.cs:59]:
```suggestion
                        Thread.Sleep(Random.Shared.Next(1, 10));
```

File: tests/Neo.Plugins.Storage.Tests/StoreTest.MultiThread.cs [hunk 2]
```diff
@@ -12,101 +12,157 @@
...
-                snapshot.Put(testKey, value);
-            });
-        }
+                for (var i = 0; i < tasks.Length; i++)
+                {
+                    var value = new byte[] { 0x04, 0x05, 0x06, (byte)i };
+                    tasks[i] = Task.Run(() =>
+                    {
+                        // Introduce delay to increase conflict chance
+                        Thread.Sleep(new Random().Next(1, 100));
```
Comment (cschuchardt88) [inline tests/Neo.Plugins.Storage.Tests/StoreTest.MultiThread.cs:108]:
```suggestion
                        Thread.Sleep(Random,Shared.Next(1, 100));
```

File: tests/Neo.Plugins.Storage.Tests/StoreTest.MultiThread.cs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    [TestMethod]
+    public void TestMultiThreadLevelDbSnapshotPutUntilException()
+    {
+        using var store = levelDbStore.GetStore(path_leveldb);
+        using var snapshot = store.GetSnapshot();
+        var testKey = new byte[] { 0x01, 0x02, 0x03 };
+
+        var threadCount = 1;
+        while (true)
```
Comment (shargon) [inline tests/Neo.Plugins.Storage.Tests/StoreTest.MultiThread.cs:97]: Mmmmm.... something more elegant?

File: src/Plugins/LevelDBStore/readme.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
+﻿Sure, here's the revised README document in a more natural, native English style:
```
Comment (vncoelho) [inline src/Plugins/LevelDBStore/readme.md:1]: Sure

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
one UT is faulted :S

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
@Jim8y one test fail

**Comment by cschuchardt88** [comment]
@Jim8y this has been fixed already. close?

**Comment by Jim8y** [comment]
fixed

**Pull request closed without merging**